fvm/src/kernel/default.rs

```diff
@@ -463,24 +464,21 @@ where
         })
     }
 
-    fn hash(&mut self, code: u64, data: &[u8]) -> Result<[u8; 32]> {
+    fn hash(&mut self, code: u64, data: &[u8]) -> Result<MultihashGeneric<32>> {
         self.call_manager
             .charge_gas(self.call_manager.price_list().on_hashing(data.len()))?;
 
-        // We only support blake2b for now, but want to support others in the future.
-        if code != BLAKE2B_256 {
-            return Err(syscall_error!(IllegalArgument; "unsupported hash code {}", code).into());
-        }
-
-        let digest = blake2b_simd::Params::new()
-            .hash_length(32)
-            .to_state()
-            .update(data)
-            .finalize()
-            .as_bytes()
-            .try_into()
-            .expect("fixed array size");
-        Ok(digest)
+        // Supported hash functions are in shared::FvmHashCode
+        let code = FvmHashCode::try_from(code).map_err(|e| {
+            if let multihash::Error::UnsupportedCode(code) = e {
+                syscall_error!(IllegalArgument; "unsupported hash code {}", code)
+            } else {
+                // will only error unsupported code, since that is all we are checking
+                // https://github.com/multiformats/rust-multihash/blob/f805963970aa509ab4e0175d296eaf5f4d8dde92/derive/src/multihash.rs#L255
+                unreachable!()
+            }
+        })?;
+        Ok(code.digest(data))
```

Comment:
Beware that you're changing the contract of this syscall! It used to return the digest, now it's returning a multihash. Is this what you intended?

Reply:
I would disagree, this doesn't change the actual syscall, but it does change how we are handling it internally. really all we care about is having owned bytes and a length, which `MultihashGeneric` provides without allocating

Reply:
Yeah, this changes the kernel, not the syscall.

Reply:
Oh yeah, I missed the patch on the syscall handler that copies only the digest over, apologies. Works for me.